File: src/bus.rs
use crate::Cartridge;
use crate::ppu::PPU;
use crate::apu::APU;
use crate::joypad::Joypad;
use crate::MAPPER;

pub struct Bus<'call> {
    work_ram: [u8; 0x0800],
    //prog_rom: Vec<u8>,
    pub ppu: PPU,
    apu: APU,
    joypad1: Joypad,
    callback: Box<dyn FnMut(&PPU, &mut Joypad) + 'call>,
}

impl<'a> Bus<'a> {
    pub fn new<'call, F: FnMut(&PPU, &mut Joypad) + 'call>(sdl_context: &sdl2::Sdl, cart: Cartridge, callback: F) -> Bus<'call> {
        Bus {
            work_ram: [0; 0x0800],
            //prog_rom: cart.prog_rom,
            ppu: PPU::new(cart.mirroring, cart.is_char_ram, cart.char_rom),
            apu: APU::new(sdl_context),
            joypad1: Joypad::new(),
            callback: Box::from(callback),
        }
    }

    pub fn read8(&mut self, addr: u16) -> u8 {
        match addr {
            0x0000..=0x1fff => self.work_ram[(addr & 0x07ff) as usize],
            0x2002 => self.ppu.read_ppustts(),
            0x2004 => self.ppu.read_oam_data(),
            0x2007 => self.ppu.read(),
            0x2008..=0x3fff => self.read8(addr & 0x2007),
            0x4015 => self.apu.read_status(),
            0x4016 => self.joypad1.read(),
            0x4017 => 0,
            0x8000..=0xffff => {
                MAPPER.lock().unwrap().read_prog_rom(addr)
                // if self.prog_rom.len() == 0x4000 {
                //     addr &= 0x3fff;
                // } else {
                //     addr &= 0x7fff;
                // }
                // self.prog_rom[addr as usize]
            }
            _ => todo!("Read from 0x{:04X} in CPU", addr),
        }
    }

    pub fn write8(&mut self, addr: u16, data: u8) {
        match addr {
            0x0000..=0x1fff => self.work_ram[(addr & 0x07ff) as usize] = data,
            0x2000 => self.ppu.write_to_ppuctrl(data),
            0x2001 => self.ppu.write_to_ppumask(data),
            0x2003 => self.ppu.write_to_oam_addr(data),
            0x2004 => self.ppu.write_to_oam_data(data),
            0x2005 => self.ppu.write_to_ppuscrl(data),
            0x2006 => self.ppu.write_to_ppuaddr(data),
            0x2007 => self.ppu.write(data),
            0x2008..=0x3fff => self.write8(addr & 0x2007, data),
            0x4000..=0x4003 | 0x4004..=0x4007 | 0x4008 | 0x400a | 0x400b | 0x400c | 0x400e | 0x400f => self.apu.write(addr, data),
            0x4014 => {
                let mut buffer: [u8; 256] = [0; 256];
                for i in 0..=255 {
                    buffer[i] = self.read8(((data as u16) << 8) + i as u16);
                }
                self.ppu.write_to_oam_dma(buffer);
                for _ in 0..513 {
                    // fixme
                    self.ppu.tick(3);
                }
            }
            0x4015 => self.apu.write_status(data),
            0x4016 => self.joypad1.write(data),
            0x4017 => self.apu.write_frame_counter(data),
            0x8000..=0xffff => {
                MAPPER.lock().unwrap().write(addr, data);
            }
            0x4010 | 0x4011 => {},
            _ => todo!("Write to 0x{:04X} in CPU", addr),
        }
    }

    pub fn poll_apu_irq(&self) -> bool {
        self.apu.irq()
    }

    pub fn tick(&mut self, cycle: usize) {
        let old_nmi = self.ppu.nmi;
        self.ppu.tick(cycle * 3);
        self.apu.tick(cycle);
        if !old_nmi && self.ppu.nmi {
            (self.callback)(&self.ppu, &mut self.joypad1);
        }
    }
}

File: source/bus.c
#include "common.h"

#define between(start, address, end) (start <= address && address <= end)

ROM *rom;
unsigned char internal_ram[0x800];

extern unsigned int cpu_cycle;
ROM *load_rom(char *file_name);
void tick_ppu(unsigned int cycle);
void init_ppu(void);
void init_apu(void);
void write_ppu_control(unsigned char value);
void write_ppu_mask(unsigned char value);
unsigned char read_ppu_status(void);
void write_oam_address(unsigned char value);
unsigned char read_oam_data(void);
void write_oam_data(unsigned char value);
void write_ppu_scroll(unsigned char value);
void write_ppu_address(unsigned char value);
unsigned char read_ppu_data(void);
void write_ppu_data(unsigned char value);
unsigned char read_joypad(void);
void write_joypad(unsigned char value);
void write_square1(unsigned short address, unsigned char value);
void write_square2(unsigned short address, unsigned char value);
void write_triangle(unsigned short address, unsigned char value);
void write_noise(unsigned short address, unsigned char value);

extern void (*init_bank)(void);
extern unsigned char (*read_bank1)(unsigned short address);
extern unsigned char (*read_bank2)(unsigned short address);
extern void (*write_bank)(unsigned short address, unsigned char value);

void tick(unsigned int cycle) {
    cpu_cycle += cycle;
    tick_ppu(cycle * 3);
}

void init_bus(char *file_name) {
    rom = load_rom(file_name);
    init_bank();
    init_ppu();
    init_apu();
}

unsigned char bus_read8(unsigned short address) {
    if(between(0x0000, address, 0x1fff)) {
        return internal_ram[address & 0x7ff];
    } else if(address == 0x2002) {
        return read_ppu_status();
    } else if(address == 0x2004) {
        return read_oam_data();
    } else if(address == 0x2007) {
        return read_ppu_data();
    } else if(between(0x2008, address, 0x3fff)) {
        return bus_read8(address & 0x2007);
    } else if(address == 0x4016) {
        return read_joypad();
    } else if(between(0x8000, address, 0xbfff)) {
        return read_bank1(address);
    } else if(between(0xc000, address, 0xffff)) {
        return read_bank2(address);
    } else if(address == 0x4017) {
        return 0;
    } else {
        error("Unsupported bus read 0x%04X\n", address);
    }
}

void bus_write8(unsigned short address, unsigned char value) {
    if(between(0x0000, address, 0x1fff)) {
        internal_ram[address & 0x7ff] = value;
    } else if(address == 0x2000) {
        write_ppu_control(value);
    } else if(address == 0x2001) {
        write_ppu_mask(value);
    } else if(address == 0x2003) {
        write_oam_address(value);
    } else if(address == 0x2004) {
        write_oam_data(value);
    } else if(address == 0x2005) {
        write_ppu_scroll(value);
    } else if(address == 0x2006) {
        write_ppu_address(value);
    } else if(address == 0x2007) {
        write_ppu_data(value);
    } else if(between(0x2008, address, 0x3fff)) {
        bus_write8(address & 0x2007, value);
    } else if(between(0x4000, address, 0x4003)) {
        write_square1(address, value);
    } else if(between(0x4004, address, 0x4007)) {
        write_square2(address, value);
    } else if(between(0x4008, address, 0x400b)) {
        write_triangle(address, value);
    } else if(between(0x400c, address, 0x400f)) {
        write_noise(address, value);
    } else if(address == 0x4014) {
        tick((cpu_cycle % 2 == 0) ? 1 : 2);
        for(int i = 0; i < 256; i++) {
            write_oam_data(bus_read8((value << 8) + i));
            tick(2);
        }
    } else if(address == 0x4016) {
        write_joypad(value);
    } else if(between(0x8000, address, 0xffff)) {
        write_bank(address, value);
    } else if(address == 0x4010 || address == 0x4011 || address == 0x4015 || address == 0x4017) {

    } else {
        error("Unsupported bus write 0x%04X\n", address);
    }
}
